### ISSUE TITLE
envoy h3 support with chrome web failed

### ISSUE THREAD

**Issue by columbus9963** [envoyproxy/envoy#24103]
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: *One line description*
I‘m using chrome client(107.0.5304.107) and quic enabled envoy proxy(v1.23.1)  for h3 web testing, but facing version negotiation error
![image](https://user-images.githubusercontent.com/57693768/203051906-0df9253c-a0f5-42fa-b2d5-e9e273207d87.png)
![image](https://user-images.githubusercontent.com/57693768/203053996-20096149-b52d-4999-870c-6a2b7fc03af8.png)

*Description*:
>Describe the issue.
does envoy really support h3-29 with google chrome client？

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.
if there has anything not clear enough，I can give the detail of the testcase above.

**Comment by danzh2010** [comment]
Envoy only support QUIC RCF v1 and we are not planning to support earlier version of QUIC. Chrome version 107.0.5304.107 is fairly new and it should support RCF v1, from the above netlog it is still hard to tell why it didn't finish the version negotiation.

**Comment by RyanTheOptimist** [comment]
Draft-29 is explicitly disabled in Envoy here:

https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/quic/platform/quiche_flags_impl.cc?L31

As Dan say, RFC v1 should be enabled by default in Chrome. Did you explicitly enable -29 with a command line flag? Envoy should also not be advertising support for -29, so I'm quite surprised that Chrome attempted a -29 handshake.

**Comment by columbus9963** [comment]
> I've been using Chrome with Envoy h3 for some time without issue. Be aware that `h3-29` is an old draft version of HTTP/3 and the standard is now an official RFC and version 1. You may need to enable QUIC in `chrome:flags`
> 
> ![image](https://user-images.githubusercontent.com/246317/203135613-82b9d663-2a95-408e-92cc-fd9834fd139b.png)
when chrome has enabled quic as above, trying h3 stream to envoy proxy still failed, the packet and error as below:
![image](https://user-images.githubusercontent.com/57693768/203197428-0188e875-a566-4467-a593-dea4ccac5b59.png)
![image](https://user-images.githubusercontent.com/57693768/203197277-4c9f0b5c-bdf0-466b-b0d3-3ab4e50eca46.png)
Meanwhile I'm using firefox client(enabled quic, version: 106.0.5 (64-bit) ) to try h3 stream with the same envoy proxy(same config file), then h3 works fine
![image](https://user-images.githubusercontent.com/57693768/203198244-844be6fb-2bf5-4c66-8417-c3b5ccd319fc.png)
![image](https://user-images.githubusercontent.com/57693768/203198899-b04ddb0c-fe83-44b1-9af8-0135f054d608.png)
Is there any difference between firefox's and chrome's h3 implementation?
Any thoughts? @danzh2010 @RyanTheOptimist

**Comment by columbus9963** [comment]
The same cert file works fine with chrome client in h2 procotol，but sent certificate unknown error in h3 tls handshake. I think there may be some gap between chrome and envoy in h3 implementation. Could you give some help for sharing a simple testcase using chrome + envoy to implement h3 web comunication? I had already tried my best to make envoy for h3 production use, but still failed with chrome.

**Comment by RyanTheOptimist** [comment]
Please collect a Chrome net log. https://www.chromium.org/for-testers/providing-network-details/

**Comment by columbus9963** [comment]
> Please collect a Chrome net log. https://www.chromium.org/for-testers/providing-network-details/

@RyanTheOptimist Thanks For your advised suggestion!

Using the network tool mention above, I saw a quic error below:
![image](https://user-images.githubusercontent.com/57693768/204453910-0f1da4db-84ba-4985-b18f-7ad07f7ec045.png)

Then digging a few bit further research, finally I found that chrome.exe provide a command line option `--origin-to-force-quic-on` to force a chrome client using quic visiting some sepical domains or hosts. For example:
```
--enable-quic  --origin-to-force-quic-on=www.facebook.com:443,static.xx.fbcdn.net:443
```

With this option enabled, H3 was enabled in chrome successfully, without any possibility to downgrade to H2 while the udp port become unavailable.
![image](https://user-images.githubusercontent.com/57693768/204455602-f4159a2c-3d50-486e-8879-85e93dedfca2.png)

**Comment by RyanTheOptimist** [comment]
Hm. You should not need to use --origin-to-force-quic-on in normal
operation. If you are able to collect a Chrome net log, I would be happy to
investigate more thoroughly.

**Comment by vsxen** [comment]
```

envoy  version: af50070ee60866874b0a9383daf9364e884ded22/1.21.1/Clean/RELEASE/BoringSSL


[2023-08-17 23:43:55.200][1089177][debug][main] [source/server/server.cc:242] flushing stats
[2023-08-17 23:43:55.200][1089177][debug][upstream] [source/common/upstream/logical_dns_cluster.cc:108] starting async DNS resolution for www.envoyproxy.io
[2023-08-17 23:43:55.200][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:60] DNS resolution for www.envoyproxy.io started
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:178] DNS resolver file event (1)
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:301] DNS for www.envoyproxy.io resolved with: flags=1073741827[MoreComing=yes, Add=yes], interface_index=0, error_code=0, hostname=teacher-glances-63085.netlify.com.
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:324] Address to add address=35.198.196.16, ttl=752
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:301] DNS for www.envoyproxy.io resolved with: flags=1073741826[MoreComing=no, Add=yes], interface_index=0, error_code=0, hostname=teacher-glances-63085.netlify.com.
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:324] Address to add address=34.124.149.177, ttl=752
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:334] DNS Resolver flushing queries pending callback
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:227] dns resolution for www.envoyproxy.io completed with status 0
[2023-08-17 23:43:55.201][1089177][debug][upstream] [source/common/upstream/logical_dns_cluster.cc:116] async DNS resolution complete for www.envoyproxy.io
[2023-08-17 23:43:55.201][1089177][debug][upstream] [source/common/upstream/logical_dns_cluster.cc:167] DNS refresh rate reset for www.envoyproxy.io, refresh rate 5000 ms
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:231] Resolution for www.envoyproxy.io completed (async)
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:148] Destroying PendingResolution for www.envoyproxy.io
[2023-08-17 23:43:55.201][1089177][debug][dns] [source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:162] DNSServiceRefDeallocate individual sd ref


[2023-08-17 23:43:55.996][1089185][debug][conn_handler] [source/server/active_tcp_listener.cc:140] [C4] new connection from 127.0.0.1:51932
[2023-08-17 23:43:56.002][1089185][debug][connection] [source/extensions/transport_sockets/tls/ssl_socket.cc:225] [C4] TLS error: 268436502:SSL routines:OPENSSL_internal:**SSLV3_ALERT_CERTIFICATE_UNKNOWN**
[2023-08-17 23:43:56.002][1089185][debug][connection] [source/common/network/connection_impl.cc:648] [C4] remote close
[2023-08-17 23:43:56.002][1089185][debug][connection] [source/common/network/connection_impl.cc:249] [C4] closing socket: 0
[2023-08-17 23:43:56.003][1089185][debug][connection] [source/extensions/transport_sockets/tls/ssl_socket.cc:225] [C4] TLS error: 268436502:SSL routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_UNKNOWN
[2023-08-17 23:43:56.003][1089185][debug][conn_handler] [source/server/active_stream_listener_base.cc:120] [C4] adding to cleanup list
[2023-08-17 23:43:56.070][1089185][debug][conn_handler] [source/server/active_tcp_listener.cc:140] [C5] new connection from 127.0.0.1:51933
[2023-08-17 23:43:56.087][1089185][debug][http2] [source/common/http/http2/codec_impl.cc:1538] [C5] updating connection-level initial window size to 268435456
[2023-08-17 23:43:56.087][1089185][debug][http] [source/common/http/conn_manager_impl.cc:322] [C5] dispatch error: Received bad client magic byte string
[2023-08-17 23:43:56.087][1089185][debug][connection] [source/common/network/connection_impl.cc:138] [C5] closing data_to_write=0 type=2
[2023-08-17 23:43:56.087][1089185][debug][connection] [source/common/network/connection_impl_base.cc:47] [C5] setting delayed close timer with timeout 1000 ms
[2023-08-17 23:43:57.091][1089185][debug][connection] [source/common/network/connection_impl_base.cc:63] [C5] triggered delayed close
[2023-08-17 23:43:57.092][1089185][debug][connection] [source/common/network/connection_impl.cc:249] [C5] closing socket: 1
[2023-08-17 23:43:57.092][1089185][debug][connection] [source/extensions/transport_sockets/tls/ssl_socket.cc:312] [C5] SSL shutdown: rc=0
[2023-08-17 23:43:57.092][1089185][debug][conn_handler] [source/server/active_stream_listener_base.cc:120] [C5] adding to cleanup list
```
[chrome-net-export-log.log](https://github.com/envoyproxy/envoy/files/12371293/chrome-net-export-log.log)

**Comment by vsxen** [comment]
[chrome-net-export-log.log](https://github.com/envoyproxy/envoy/files/12371682/chrome-net-export-log.log)
